mlflow/pytorch/__init__.py

```diff
@@ -287,6 +288,93 @@ def gen_data():
     )
 
 
+@experimental
+def log_state_dict(state_dict, artifact_path, **kwargs):
+    """
+    Log a state dict as an MLflow artifact for the current run.
+
+    :param state_dict: PyTorch state dict to be saved.
+    :param artifact_path: Run-relative artifact path.
+    :param kwargs: kwargs to pass to ``torch.save`` method.
+
+    .. code-block:: python
+        :caption: Example
+
+        import torch
+        import mlflow
+        import mlflow.pytorch
+        # X data
+        x_data = torch.Tensor([[1.0], [2.0], [3.0]])
+        # Y data with its expected value: labels
+        y_data = torch.Tensor([[2.0], [4.0], [6.0]])
+        # Partial Model example modified from Sung Kim
+        # https://github.com/hunkim/PyTorchZeroToAll
+        class Model(torch.nn.Module):
+            def __init__(self):
+               super().__init__()
+               self.linear = torch.nn.Linear(1, 1)  # One in and one out
+            def forward(self, x):
+                y_pred = self.linear(x)
+            return y_pred
+        # our model
+        model = Model()
+        criterion = torch.nn.MSELoss(size_average=False)
+        optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
+        # Training loop
+        for epoch in range(500):
+            # Forward pass: Compute predicted y by passing x to the model
+            y_pred = model(x_data)
+            # Compute and print loss
+            loss = criterion(y_pred, y_data)
+            print(epoch, loss.data.item())
+            #Zero gradients, perform a backward pass, and update the weights.
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+        # After training
+        for hv in [4.0, 5.0, 6.0]:
+            hour_var = torch.Tensor([[hv]])
+            y_pred = model(hour_var)
+            print("predict (after training)",  hv, model(hour_var).data[0][0])
+        # log the model
+        with mlflow.start_run() as run:
+            mlflow.log_param("epochs", 500)
+            mlflow.pytorch.log_state_dict(model, "models")
+
+            # logging scripted module
+            scripted_pytorch_model = torch.jit.script(model)
+            mlflow.pytorch.log_state_dict(scripted_pytorch_model, "models")
+    """
+
+    with TempDir() as tmp:
+        local_path = tmp.path("model")
+        save_state_dict(state_dict=state_dict, path=local_path, **kwargs)
+        mlflow.tracking.fluent.log_artifacts(local_path, artifact_path)
+
+
+@experimental
+def save_state_dict(state_dict, path, **kwargs):
+    """
+    Save state dict to a path on the local file system
+
+    :param state_dict: state dict of the pytorch model. Ex: model.state_dict()
+    :param path: Local path where the state_dict is to be saved.
+    :param kwargs: kwargs to pass to ``torch.save`` method.
+    """
+
+    import torch
+
+    os.makedirs(path)
+
+    model_data_subpath = "data"
```

Comment:
Do we need this `data` directory?

Reply:
https://github.com/mlflow/mlflow/blob/9991104e157a30ae50df295d792291254ed21b96/mlflow/pytorch/__init__.py#L464 - In save model, the actual model is stored under `data` folder. I feel it would be better to follow the same convention. In future, when state dict is considered as model, the folder structure would be uniform.

Reply:
If we use the same folder structure, it's a bit confusing isn't it?

```python
# This folder structure looks like an MLflow PyTorch model, but actually is NOT.
- model
  - data
    - state_dict.pt

# How the folder structure would look like if we supported logging a state_dict as an MLflow model
- model
  - MLmodel
  - conda.yml 
  - data
    - state_dict.pt
```

Reply:
Can you please let me know your recommendation ?

should we save it like below ?

```
- model
  - state_dict.pt
```

Reply:
In future, when the state dict implementation is considered as model, We don't have to alter the structure again. And we can reuse the code from `load_model` if the same structure is followed. But it is just my opinion, if you feel otherwise, we can change it.

Reply:
I think we'll update `log_model`.

Reply:
Okie. Should we remove the `data` folder then ?

Reply:
Yes, and another thing we need to consider is which one the following usages makes more sense for users.


```python
# specify a directory
log_state_dict(state_dict, "dir1")
log_state_dict(state_dict, "dir2")

load_state_dict("dir1")
load_state_dict("dir2")

- dir1
  - state_dict.pt
- dir2
  - state_dict.pt

# ------------- vs. -------------

# specify a file path
log_state_dict(state_dict, "dir/state_dict1.pt")
log_state_dict(state_dict, "dir/state_dict2.pt")

load_state_dict("dir/state_dict1.pt")
load_state_dict("dir/state_dict2.pt")

- dir
  - state_dict1.pt
  - state_dict2.pt
```

Reply:
I have removed `data` folder .

Reply:
The seconds approach would be too confusing for users from pytorch side. When user wants to log the model, they would need to give directory as input and while using state dict they need to provide filename as input. 

The first approach(the one we follownow ) is uniform and would avoid too many confusions for the user.